src/analysis/souffle/sensitivity_analysis.dl

```diff
@@ -1,90 +1,134 @@
+
+.decl isAggregationOperator(op: Operator)
+isAggregationOperator(SQL_SUM_OPERATOR).
+isAggregationOperator(SQL_AVERAGE_OPERATOR).
+isAggregationOperator(SQL_COUNT_OPERATOR).
+isAggregationOperator(SQL_MAX_OPERATOR).
+// Aggregation operations point to where noise needs to be added. 
+.decl isAggregationOperation(sql_operation: Operation)
+isAggregationOperation(operation) :- isSqlOperation(operation), 
+      operationHasOperator(operation, operator),
+      isAggregationOperator(operator).
 
 .decl isSqlValue(value: AccessPath)
 isSqlValue(value) :- isSqlOperation(operation),
       operationHasResult(operation, value).
 
-// keeping track of all initial inputs as they all have sensitivity score 1
-.decl isInitialInput(ingress_point: AccessPath)
+// Keeping track of all initial inputs. 
+.decl isInitialInput (ingress_point: AccessPath)
+isInitialInput(as(ingress_point, AccessPath)) :- isOperation(operation), 
+      operationHasOperator(operation, CORE_INPUT_OPERATOR), 
+      operationHasResult(operation, ingress_point), 
+      operationHasAttribute(operation, [_, $StringAttributePayload(_)]).
 
-// helper functions to flatten input lists
+// Helper functions to flatten input/operand lists. 
 .decl intermediateFlatten(list:OperandList, sublist:OperandList, list_element:AccessPath)
 
-intermediateFlatten(list, sublist, list_element) :- isSqlOperation(operation), 
-      operationHasOperandList(operation, list), 
-      list = [list_element, sublist].
-intermediateFlatten(list, sublist, list_element) :- intermediateFlatten(list, [list_element, sublist], _).
+intermediateFlatten(list, sublist, list_element) :- isOperation(operation), 
+      operationHasOperandList(operation, list),
+      operationHasOperandListSuffix(operation, [list_element, sublist]).
 
 .decl flatten(list:OperandList, list_element:AccessPath)
 flatten(list, list_element) :- intermediateFlatten(list, _, list_element).
 
-// flattening each individual input per value
+// Flattening each individual input per value. 
+
 .decl isSqlInput(operation: Operation, sql_input: AccessPath)
 
 isSqlInput(operation, sql_input) :- isSqlOperation(operation), 
       operationHasOperandList(operation, input_list), 
       flatten(input_list, sql_input). 
 
-// keeping track of sensitivity score per each input at index 
-// used to calculate the maximum input sensitivity 
-// decomposed into base and inductive cases because recursive aggregates are not supported yet 
-// for the compiler case in souffle
-.decl sqlOperationHasSensitivityScoreAtIndex(operation: Operation, sensitivity_score: number, index: number) 
+// Keeping track of sensitivity score per each input at index. 
+// This is used to calculate the maximum input sensitivity. 
+// It's decomposed into base and inductive cases because recursive 
+// aggregates are not supported yet for the compiler case in souffle
+.decl sqlOperationHasSensitivityScoreAtIndex(operation: Operation, sensitivity_score: SensitvityScore, index: number) 
```

Comment:
(This comment is also based on the definition of the relation below).

I think you need some relation that gives the sensitivity of **operand**s  (AccessPaths) rather than operations, so it should be more like: `.decl  pathHasSensitivity(operand: AccessPath, sensitivity: SensitivityScore)`

and the sensitivity of the path should depend on the _operation_ that defines this path. It should be the function sensitivity of the operation in the definition multiplied by the maximum over the sensitivities of the operations in the inputs. You might need an auxiliary definition to compute the maximum over the list of inputs. So the following should almost work, but it might not be stratifiable:
```
.decl maxSensitivityInOperandList(operand_list: OperandList, sensitivity_score: SensitivityScore)
maxSensitivityInOperandList(nil, 0).
maxSensitivityInOperandList([path, subList], max(pathSensitivity, listSensitivity)) :- 
        isOperandList(subList),
        pathHasSensitivity(path, pathSensitivity),
        maxSensitivityInOperandList(subList, listSensitivity).

pathHasSensitivity(result_path, sensitivity) :- 
        operationHasResult(result_path, operation),
        operationHasOperandList(operation, input_list), 
        operationHasOperator(operation, operator),
        sqlFunctionSensitivity(operator, function_sensitivity),
        maxSensitivityInOperandList(input_list, max_input_sensitivity),
        sensitivity = function_sensitivity * max_input_sensitivity
```

To make this stratifiable we might need to add the index and do induction on indexes as you are doing, or maybe another way is to merge the functionality of `pathSensitivity` and `maxSensitivityInOperandList` somehow...

Reply:
OK my mistake, when reading this code, the drop-down arrow was hiding the rest of the definition of `sqlValueSensitivity`, which does look right to me! I should have noticed the arrow was hiding it...

Reply:
Maybe it's also useful to explain that the sensitivity of the operation is used as a proxy for the sensitivity of the access path defined by this operation.

Reply:
Done.